### PR TITLE
Allow `apply_ufunc` to ignore missing core dims

### LINKED ISSUE #6391: apply_ufunc and Datasets with variables without the core dimension
### Is your feature request related to a problem?

Consider this example 

```python
ds = xr.Dataset({"a": ("x", [1, 2, 3]), "b": ("y", [1, 2, 3])})
xr.apply_ufunc(np.mean, ds, input_core_dims=[["x"]])
```

This raises 
```
ValueError: operand to apply_ufunc has required core dimensions ['x'], but some of these dimensions are absent on an input variable: ['x']
```

because core dimension `x` is missing on variable `b`. This behaviour makes it annoying to use `apply_ufunc` on Datasets.

### Describe the solution you'd like

Add a new kwarg to `apply_ufunc` called `missing_core_dim` that controls how to handle variables without *all* input core dimensions. This kwarg could take one of two values:
1. `"raise"` - raise an error, current behaviour
2. `"copy"` - skip applying the function and copy the variable from input to output.
3. `"drop"`- skip applying the function and drop the variable.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] Closes #6391
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

This probably needs a review and another turn, maybe some more tests on multiple objects etc. There are a couple of questions inline.

### CODE REVIEW

**Comment by max-sixty** [comment]
Any thoughts on this, team?

File: xarray/core/computation.py [hunk 1]
```diff
@@ -408,7 +416,26 @@ def apply_dict_of_variables_vfunc(
 
     result_vars = {}
     for name, variable_args in zip(names, grouped_by_name):
-        result_vars[name] = func(*variable_args)
+        # TODO: is this a reasonable to check for missing core dims? We check for a dims
+        # property to protect against the case where a numpy array is passed in.
+        if hasattr(variable_args[0], "dims") and set(
+            signature.all_input_core_dims
+        ) - set(variable_args[0].dims):
```
Comment (mathause) [inline xarray/core/computation.py:423]: Do we need to check all variable_args here?

File: xarray/core/computation.py [hunk 2]
```diff
@@ -408,7 +416,26 @@ def apply_dict_of_variables_vfunc(
...
+            if missing_core_dim == "raise":
+                raise ValueError(
+                    f"Missing core dimension on {name!r}. Either add the core dimension, or set `missing_core_dim` to `copy` or `drop`."
+                )
+            elif missing_core_dim == "copy":
+                # TODO: is it correct to copy the first variable here?
+                result_vars[name] = variable_args[0]
```
Comment (mathause) [inline xarray/core/computation.py:430]: Sounds reasonable to me.

**Review by mathause — COMMENTED** [review]
We may have to think about how we handle multiple input_vars and what happens if multiple `ds` are passed (e.g `apply_ufunc(func, ds1, ds2)`). But that are probably rather edge-cases.

**Comment by max-sixty** [comment]
What's the best name for this arg? 
- `missing_core_dim`? Possibly too synonymous with other args which actually list dims, like `exclude_dims`?
- `on_missing_core_dim`? Arguably more fluent with `on_missing_core_dim="raise"`, but less standard...

File: xarray/core/computation.py (hunk 1 above)
Comment (max-sixty) [reply xarray/core/computation.py]: Yes! Now done

**Comment by max-sixty** [comment]
OK, I'm reasonably happy with this now — it's ready to go, pending docs & changelog. Any thoughts prior?

**Comment by max-sixty** [comment]
> We may have to think about how we handle multiple input_vars and what happens if multiple `ds` are passed (e.g `apply_ufunc(func, ds1, ds2)`). But that are probably rather edge-cases.

Sorry, just saw this. Yes, I actually had a TODO prior re adding tests for this, which I just added.

In the case of `drop` & `raise`, the order doesn't matter. In the case of `copy`, we copy from the first argument. Do you think that's reasonable?

File: xarray/core/computation.py [hunk 3]
```diff
@@ -396,8 +399,31 @@ def _unpack_dict_tuples(
...
+        # Check whether all the dims are on the variable. Note that we need the
+        # `hasattr` to check for a dims property, to protect against the case where
+        # a numpy array is passed in.
+        if hasattr(variable_arg, "dims") and set(core_dims) - set(variable_arg.dims):
+            # Slightly awkward design, of returning the error message. But we want to
+            # give a detailed error message, which requires inspecting the variable in
+            # the inner loop.
+            return (
+                f"Missing core dimension(s) {set(core_dims) - set(variable_arg.dims)} on `{name}`. "
+                "Either add the core dimension, or set `missing_core_dim` to `copy` or `drop`. "
```
Comment (dcherian) [inline xarray/core/computation.py:413]:
```suggestion
                "Either add the core dimension using `broadcast` or `expand_dims` for example, or set `missing_core_dim` to `copy` or `drop`. "
```

File: xarray/core/computation.py [hunk 4]
```diff
@@ -408,7 +434,20 @@ def apply_dict_of_variables_vfunc(
 
     result_vars = {}
     for name, variable_args in zip(names, grouped_by_name):
-        result_vars[name] = func(*variable_args)
+        core_dim_check = _check_core_dims(signature, variable_args, name)
+        if core_dim_check is True:
```
Comment (dcherian) [inline xarray/core/computation.py:438]: nit: 

```suggestion
        core_dim_present = _check_core_dims(signature, variable_args, name)
        if core_dim_present is True:
```

but feel free to ignore if you don't like it.

File: xarray/core/computation.py (hunk 3 above)
Comment (dcherian) [reply xarray/core/computation.py:413]: Can we raise one error listing all variables with their respective missing core dimensions?

File: xarray/core/computation.py [hunk 5]
```diff
@@ -408,7 +434,20 @@ def apply_dict_of_variables_vfunc(
...
+            result_vars[name] = func(*variable_args)
+        else:
+            if missing_core_dim == "raise":
+                raise ValueError(core_dim_check)
+            elif missing_core_dim == "copy":
+                result_vars[name] = variable_args[0]
+            elif missing_core_dim == "drop":
+                pass
+            else:
+                raise ValueError(
```
Comment (dcherian) [inline xarray/core/computation.py:448]: We should check to allow only`"raise"` is the only valid option for dataarrays as input.

File: xarray/core/computation.py [hunk 6]
```diff
@@ -851,6 +888,7 @@ def apply_ufunc(
     output_sizes: Mapping[Any, int] | None = None,
     meta: Any = None,
     dask_gufunc_kwargs: dict[str, Any] | None = None,
+    missing_core_dim: MissingCoreDimOptions = "raise",
```
Comment (dcherian) [inline xarray/core/computation.py:891]: Needs addition to docstring.

File: xarray/core/computation.py (hunk 4 above)
Comment (max-sixty) [reply xarray/core/computation.py:438]: Better, thanks!

File: xarray/core/computation.py (hunk 3 above)
Comment (max-sixty) [reply xarray/core/computation.py:413]: Cool, latest version has this as the message (repeating for more vars):

```
E                   ValueError: Missing core dims from x_z on variable(s):
E
E                   Missing {'y'} on
E
E                   <xarray.Variable (x: 2, z: 3)>
E                   array([[0, 1, 2],
E                          [3, 4, 5]])
E
E                   Either add the core dimension, or if passing a dataset alternatively pass `on_missing_core_dim` as `copy` or `drop`.
```

File: xarray/core/computation.py (hunk 5 above)
Comment (max-sixty) [reply xarray/core/computation.py:454]: Actually this code path only gets hit for Datasets — do we want to pre-emptively check for the kwarg when there are only DataArrays?

Comment (dcherian) [reply xarray/core/computation.py:454]: Yes I think so.

File: xarray/core/computation.py [hunk 7]
```diff
@@ -396,8 +399,37 @@ def _unpack_dict_tuples(
...
+    """
+    missing = []
+    for core_dims, variable_arg in zip(signature.input_core_dims, variable_args):
+        # Check whether all the dims are on the variable. Note that we need the
+        # `hasattr` to check for a dims property, to protect against the case where
+        # a numpy array is passed in.
+        if hasattr(variable_arg, "dims") and set(core_dims) - set(variable_arg.dims):
+            missing += [[variable_arg, core_dims]]
+    if missing:
+        message = f"Missing core dims from `{name}` on variable(s):\n\n"
```
Comment (dcherian) [inline xarray/core/computation.py:418]: Now I'm a bit confused, is this wording correct/better?

```suggestion
        message = f"Missing core dims on some variables named `{name}`:\n\n"
```

File: xarray/core/computation.py [hunk 8]
```diff
@@ -396,8 +399,37 @@ def _unpack_dict_tuples(
...
+        for variable_arg, core_dims in missing:
+            message += f"Missing {set(core_dims) - set(variable_arg.dims)} on\n{variable_arg}\n\n"
```
Comment (dcherian) [inline xarray/core/computation.py:420]: This may be clearer if we could add an "argument number".

So "Missing ['x'] in argument number 2"

File: xarray/core/computation.py (hunk 7 above)
Comment (max-sixty) [reply xarray/core/computation.py:418]: It is indeed...

File: xarray/core/computation.py (hunk 5 above)
Comment (max-sixty) [reply xarray/core/computation.py:454]: I'm in gentle disagreement — if someone passes `apply_ufunc(f, da, on_missing_core_dims="copy")` then the claim is that we want to raise immediately?

I think that's going to lead to lots of callers having to check the type and choose what to pass based on whether it's a `da` vs. `ds`. The motivating example from https://github.com/pydata/xarray/pull/8114 would need to do this.

Instead, surely it's consistent to allow `drop` & `copy` for arrays, but know they'll get ignored?

File: xarray/core/computation.py (hunk 8 above)
Comment (max-sixty) [reply xarray/core/computation.py]: OK I rewrote it from scratch after some sleep, WDYT? Please continue refining, your suggestions have been excellent

(These error messages take time but I actually think it's really nice when they're nice — as someone returning to xarray as being occasionally befuddled. A good exercise in usability!)

Comment (max-sixty) [reply xarray/core/computation.py]: Here's the full error message:

```
E                   ValueError: Missing core dims {'y'} from arg number 2 on a variable named `x_z`:
E                   <xarray.Variable (x: 2, z: 3)>
E                   array([[0, 1, 2],
E                          [3, 4, 5]])
E
E                   Either add the core dimension, or if passing a dataset alternatively pass `on_missing_core_dim` as `copy` or `drop`.
```

(Would love a pytest-accept-like solution to snapshot these, much easier to both iterate on and understand as a reviewer than a partial regex)

File: xarray/core/computation.py (hunk 5 above)
Comment (dcherian) [reply xarray/core/computation.py:456]: > think that's going to lead to lots of callers having to check the type and choose what to pass based on whether it's a da vs. ds.

Ok good point. I'm convinced!

File: xarray/core/computation.py (hunk 8 above)
Comment (dcherian) [reply xarray/core/computation.py]: Perhaps a slight reordering? This is quite minor so feel free to ignore

"Missing core dims {'y'} on a variable named `x_z` in arg number 2"
